Integrate TaskManager into NodeGraph and Discovery

tegefaulkes opened the pull request:
### Description

This PR focuses on updating `Nodes`, `Discovery` domain to use the new `Tasks` system.

Generally there are 3 places where background queues are being used. Each of these need to be updated to use the new tasks system from #438.

1. Discovery system
2. Nodes pinging and authenticating nodes.
3. Nodes refreshing buckets on a timeout.
4. `NodeConnectionManager`  Does the network entry procedure with `syncNodeGraph`.

#### SetNode details
In `NodeConnectionManager` when adding a node to the `NodeGraph` with `nodeManager.setNode` we can end up with the case where a bucket is full. When this happens we need to ping nodes within the bucket to determine if they're still alive and remove any nodes that don't respond so we can add the new one. We need to convert this to use the scheduler/queue.

By default the `nodeManager.setNode` doesn't ping a node to check if it's online before adding it. It is expected that you ping the node before using `setNode` to add it. We add nodes whenever we interact or discover a node. This can happen in the cases of... We learn about a node from other nodes, the a connection to the learned node hasn't been made so it needs to be pinged. A node has connected to us so it just needs to be added. We connected to a node, details need to be updated. Given that we needed the ability to ping and `setNode` or just `setNode`. 

#### RefreshBuckets details
`nodeManager.refreshBucket` needs to updated to use the new scheduler/queue system. In this case we will be making use of the `Scheduler` features. A refresh bucket operation needs to be run on a bucket if the bucket hasn't seen activity for an hour. Given this we need schedule each bucket for an hour delay. If a bucket is updated we need to reset the timer for the bucket. To do this we need to make use of the `taskPath` and timer updating features of the `Scheduler`.  These `refreshBucket` tasks should be the lowest priority.

Having `refreshBuckets` system use the `Tasks` system is a little complex since it works on a kind of watchdog system. here are the requirements of the `refreshBucket` system
1. A `refreshBucket` operation selects a random `NodeId` within the target bucket's range of nodes and preforms a search for that node.
2. Each bucket needs to be scheduled to do a `refreshBucket` operation every hour.
3. If data is updated within a bucket then this scheduled delay is reset to an hour.

here are some relevant constraints of the tasks system.
1. A task is scheduled with a delay and is removed when cancelled or completed.
2. A task state can be checked. It can be `scheduled`, `queued`, `active`, `success` or `failure`.
3. A scheduled task's delay can be updated if it is in the `scheduled` state.
4. A task can be 'grouped' using the task's `path`, We can track a task for a single bucket using a path such as `['refreshBucket', bucketIndex]`. This can enable us to find existing tasks for buckets.

Given these constraints I think we need to do the following.
1. During start we need to iterate over the existing `refreshBucket` tasks using the `tasks.getTasksByPath`, reset the delay on existing tasks and create ones for buckets missing them.
2. When any bucket get updated using `nodeManager.setNode` then we need to update the delay of that refresh bucket task. This can be done by getting the task for the bucket using `tasks.getTasksbyPath` and updating the scheduled delay if the task is in the `scheduled` state. If it is `queued` or `Active` then ignore, if not task exist create one. The tasks's delay can be updated by either canceling it and re-creating or using a provided `updateDelay` method.
3. The tasks themselves are technically emphemiral so we could remove all `refreshBuckets` tasks during `nodeManager.stop()`.  Minor detail, won't really make a difference to operation.

#### NodeConenctionManager details
In the `NodeConnectionManager` we have a method `syncNodeGraph` that does the following proceedure. If `syncNodeGrap` is run with blocking = false then we need to run it in the background. 
1. for each seed node we request the closest nodes to our `NodeId`. Then go through this list pinging and adding their details to our nodeGraph.
2. Then for every bucket above the closet node's bucket we run a `refreshBucket` operation.

Looking over this I think it should be part of `NodeManager` and making calls to `NodeConnectionManager` for pinging.

As for how to implement this using `Tasks`.
1. Do an initial getRemoteNodeClosestNodes to the seed node. This is blocking but it could also be a task.
2. The first step then schedules `pingAndAdd` tasks for each of the closest nodes.
3. The first step then schedules with 0 delay a `refreshBucket` operation for every bucket above the closest node's bucket.

#### Other details
Other aspects to consider. The Kademlia `findNode` operation in the `NodeConnectionManager` is considered a single operation but is in essence a priority queue search for the target node. We can consider splitting this up into a compound task where each step in the search can be it's own task within the process. This would apply to the `refreshBucket` operation since that is doing a `findNode` as well.

Handlers will need to support cancel-ability. They must take an abort signal and quickly end operation.

### Issues Fixed

* Fixes #418
* Fixes #415
* Fixes #414
* Fixes #413
* Fixes #307
* Related #353 - This will be a separate PR after this one. After we address updating `Discovery` to use `TaskManager`.
* Related #328 
* Related #329 - this issue contains notes on how to refactor the discovery and prioritising gestalt vertices
* Related #297 - this issue contains notes on what should be made timed/cancellable

### Tasks
 - [x] 1. Add`TaskManager` to `PolykeyAgent`
   - At the end of PolykeyAgent.start you need to do scheduler.startProcessing(), it should be a lazy creation and lazy start from the beginning to allow handlers to be registered. This is why PolykeyAgent would be creating a lazy scheduler so it doesn't start the processing loop.
 - [x] 2. Update nodes pinging/authentication queue to use the new tasks system
 - [x] 3. Update refresh buckets to use new tasks system.
 - ~4. Update `Discovery` domain to use the new tasks system.~
 - [x] 5. Update and check nodes tests.
 - ~6. Update and check `Discovery` tests.~
 - ~7. Address issue #353~ 
 - [x] 8. Address issue #418
 - [x] 9. Address issue #415
 - ~10. Address issue #414~ - Needs more testing, can't address this now. I've added #307 to help with part of this.
 - [x] 11. Address issue #413
 - [x] 12. refactor `setNode` garbage collection https://github.com/MatrixAI/Polykey/pull/445#issuecomment-1246175185
 - [x] 13. Address issue #307  

### Final checklist
<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

* [x] Domain specific tests
* [x] Full tests
* [x] Updated inline-comment documentation
* [x] Lint fixed
* [x] Squash and rebased
* [x] Sanity check the final build

tegefaulkes committed:
feat: `PolykeyAgent.ts` using `TaskManager`

tegefaulkes committed:
feat: updated `NodeManager` to use `TaskManager`

tegefaulkes committed:
feat: updated `NodeConnectionManager` to use `TaskManager`

CMCDragonkai commented on src/nodes/NodeConnectionManager.ts:
```diff
@@ -605,46 +631,54 @@ class NodeConnectionManager {
-          } else {
+          const pingAndSetTask = await this.taskManager.scheduleTask({
+            delay: 0,
+            handlerId: this.pingAndSetNodeHandlerId,
```
Suggest using arrow function property on the class for this. Then the name of the property could be the this handler ID too?

CMCDragonkai replied:
```
[this.constructor.name, this.pingAndSetNodeHandler.name].join('.')
// NodeConnectionManager.pingAndSetNodeHandler
```

We would just be careful about this if we ever do a schema migration...

And schema migration is relevant upon doing a 1.0 release.

If a class name/handler name changes, then a schema migration will be necessary as well.

We can code in some "constraints" into our `schema` domain later to see if it DB's schema matches up with what is expected... But we do have a place to code up forward migrations.

tegefaulkes replied:
I have addressed this.

tegefaulkes committed:
feat: adding cancellability to `NodeManager` handlers

tegefaulkes committed:
fix: small bug with `pingNode`

tegefaulkes committed:
fix: bugs with `nodeConnectionManager.syncNodeGraph`

tegefaulkes committed:
tests: cleaning up dependencies in tests

tegefaulkes committed:
fix: `getRemoteNodeClosestNodes` shouldn't throw connection errors

`getRemoteNodeClosestNodes` was throwing an connection error in certain conditions. If it failed to connect to a node it should've just skipped that node.

#418

tegefaulkes committed:
fix: excessive connections from `refreshBuckets`

Removing excessive logging for using connections. We don't need a 3 log messages for each time we use an existing connection.

Adding 'jitter' or spacing to the `refreshBuckets` delays so that they don't run all at once. This is implemented with a `refreshBucketDelaySpread` paramater that specifies the multiple of the delay to spread across. defaults to 0.5 for 50%

Adding a 'heuristic' to `refreshBucket` to prevent it from contacting the same nodes repeatably. Currently this is just a check in `getClosestGlobalNodes` where if we find less than `nodeBucketLimit` nodes we just reset the timer on all `refreshBucket` tasks.

Adding tests for checking the spread of `refreshBucket` delays. Another test for resetting the timer on `refreshBucket` tasks if a `findNode` finds less than 20 nodes.

#415

tegefaulkes committed:
feat: `nodeConnectionManager.getClosestGlobalNodes` can optionally skip recently offline nodes

This is done with an in-memory map of `nodeIdstring` to some data tracking the backoff period. it defaults to 5 min and doubles each failure.

#413

tegefaulkes committed:
feat: added handler to detect promise deadlocks to `ExitHandlers.ts`

#307

tegefaulkes committed:
fix: handerIds are derived from class and handler function names

CMCDragonkai commented on src/nodes/NodeManager.ts:
```diff
@@ -410,6 +459,7 @@ class NodeManager {
     force: boolean = false,
     timeout?: number,
     tran?: DBTransaction,
+    options: { signal?: AbortSignal } = {},
```
The `timeout` and the `signal` should be replaced with a `ContextTimed` object.

Use the `timedCancellable` decorator here.

This is a bit more complicated than original design.

The reason is that the `setNode` can schedule background tasks or not.

In the case of `block: false`, then everything runs within the same promise lifecycle. In which case the `timer` and the `signal` gets propagated down to all the downstream functions. Now this would mean:

```ts
public setNode(
  nodeId: NodeId,
  nodeAddress: NodeAddress,
  block?: boolean,
  force?: boolean,
  ctx?: Partial<ContextTimed>
): PromiseCancellable<void>;
@ready(new nodesErrors.ErrorNodeManagerNotRunning())
@timedCancellable(false, 1000)
public async setNode(
  nodeId: NodeId,
  nodeAddress: NodeAddress,
  block: boolean = true,
  force: boolean = false,
  @context ctx: ContextTimed
): Promise<void> {

  // ...

}
```

CMCDragonkai replied:
Once this change takes place. The meaning of `timeout` is different. The original `timeout` corresponds to the connection establishment timeout that occurs for every proxy connection.

However in this case, that's not what what the timer means. Instead timer is now defaulted to 1000 ms. Which is the total amount of time allowed to complete a `setNode` operation.

CMCDragonkai replied:
The `ctx.timer.getTimeout()` can then be used to pass into `pingNode` which has independent timeouts that is all occurring in parallel. This works for the `block: true`.

What about `block: false`, in this case, we would propagate the `ctx.timer.delay` into the `TaskDeadline`, which itself is then propagated to the `pingNode` running inside the task handler function.

CMCDragonkai replied:
This applies to the intermediate `garbageCollectOldNode` too.

In the new garbage collection mechanism where we are maintaining pending nodes. We would be iterating the existing nodes in a bucket in parallel and pinging them. However we may wish still abide by the k-parallel constant.

This `k parallel constant` should be written to the class as a parameter.

tegefaulkes replied:
I've converted functions called inside of the task handlers to be a `timedCancellable`. So far it's working fine. More work needs to be done  with converting methods to `timedCancellable` but I think this should be left to a separate PR that focuses on implementing deadlines and cancellablility where needed.

tegefaulkes replied:
The `timedCancellable` conversions are done fr this PR. `setNode` was updated to be a `timedCancellable` but the timer is for the overall task completeion. The `pingNode`s as part of the garbage collection are started with their own timers set by the `pingTimeout` parameter.

Further `timedCancellable` conversion will be done in a later PR.

CMCDragonkai commented on src/nodes/NodeManager.ts:
```diff
@@ -474,45 +536,48 @@ class NodeManager {
     nodeAddress: NodeAddress,
     timeout?: number,
+    options: { signal?: AbortSignal } = {},
   ) {
```
The garbage collection mechanism is changing to become like a while loop that goes over all the existing and pending nodes. We will need a `Semaphore` exposed in the `async-locks` library so the while loop can iterate over the combined bucket upon acquiring a lock. The semaphore limit is set to the k parallel constant.

This should allow us to collapse the existing and pending nodes into a single bucket with the remaining nodes.

CMCDragonkai replied:
The maximum number of pings that could occur is the 20 bucket limit, but the minimum is however many nodes there are in the pending bucket partition multiplied by the k parallel constant.

CMCDragonkai replied:
Actually max is 20, min will be 3. The true number of pings will be somewhere in between.

tegefaulkes replied:
Garbage collection has been updated. As a result, `setNode` is no longer has optional blocking or abortion with signal.

CMCDragonkai commented on src/nodes/NodeManager.ts:
```diff
@@ -573,169 +641,153 @@ class NodeManager {
-      // Bucket is same as next bucket, this affects the timer
-      this.refreshBucketRefreshTimer();
+  @ready(new nodesErrors.ErrorNodeManagerNotRunning())
+  public async updateRefreshBucketDelay(
```
In the process of refreshing the bucket delay, the `getTask` and `updateTask` cannot share the same transaction context.

```ts
const task = await getTask();
try {
  await updateTask(task.id);
} catch (e) {
  if (e instanceof ErrorTaskMissing) {
     // create it
  } else if (e instanceof ErrorTaskRunning) {
    // If it is already running
    // Do nothing!
  }
}
```

This is because if the `getTask` and `updateTask` shared a context, even though `updateTask` is locking the `TaskId`, the snapshot is created at `getTask`, and 2 transactions doing a get, then put, can have a conflict. We don't want a conflict here because we don't expect a conflict exception due to task mutation when interacting with the nodes domain.

So here, we are relying on the exceptions of `updateTask` to know whether we can do what we want.

tegefaulkes replied:
This has been fixed.

CMCDragonkai commented on src/nodes/NodeManager.ts:
```diff
@@ -573,169 +641,153 @@ class NodeManager {
+        const spread =
+          (Math.random() - 0.5) *
+          this.refreshBucketDelay *
+          this.refreshBucketDelaySpread;
```
Can this be a utility function. Call it `nodesUtils.refreshBucketsDelayJitter`. It should a function that can give you this number that you adding to the `delay`.

tegefaulkes replied:
I have extracted this into the utils.

CMCDragonkai commented on src/nodes/NodeManager.ts:
```diff
@@ -36,37 +36,67 @@ class NodeManager {
+      handlerId: this.refreshBucketHandlerId,
+      lazy: true,
+      parameters: [bucketIndex],
+      path: ['refreshBucket', `${bucketIndex}`],
```
This should become a class property. It should also have the class name as the prefix. Similar to the task handler name as well. That means using `this.constructor.name`.

tegefaulkes replied:
This has been addressed for all handlers.

CMCDragonkai commented on tests/client/service/nodesAdd.test.ts:
```diff
@@ -128,16 +130,17 @@ describe('nodesAdd', () => {
     });
   });
   afterEach(async () => {
+    await taskManager.startProcessing();
```
This should be `stopProcessing()`. Check all the rest as well.

tegefaulkes replied:
This has been checked for all tests and fixed.

tegefaulkes committed:
fix: fixing up `setNode` garbage collection.

tegefaulkes committed:
fix: refactored `nodeManager.setNode` garbage collection

tegefaulkes committed:
tests: proper stopping of `taskManager` in tests

tegefaulkes committed:
fix: updating task paths

`TaskPaths` should take the form of `[basePath, handlerId, ...extra]`. basePath is the `this.constructor.name` for the domain the handler is registered for.

CMCDragonkai commented on src/nodes/NodeGraph.ts:
```diff
@@ -151,6 +151,15 @@ class NodeGraph {
   }
 
+  @ready(new nodesErrors.ErrorNodeGraphNotRunning())
+  public async lockBucket(bucketIndex: number, tran: DBTransaction) {
```
Should this be `protected`?

tegefaulkes replied:
It's getting used inside of `NodeManager` to lock the bucket during `setNode` and `garbageCollectBucket`.

CMCDragonkai commented on src/tasks/TaskManager.ts:
```diff
@@ -1148,6 +1148,7 @@ class TaskManager {
     this.logger.debug(`Requeued Task ${taskIdEncoded}`);
   }
 
+  @ready(new tasksErrors.ErrorTaskManagerNotRunning())
```
This should not be needed. It is a protected function so you don't need a `ready` decorator.

tegefaulkes replied:
It's protected here but it's callable at any time with `task.cancel()` or `task.promise().cancel()`. I added it while debugging but it's not really needed.

tegefaulkes replied:
I've removed it.

CMCDragonkai commented on src/nodes/utils.ts:
```diff
@@ -310,6 +313,14 @@ function generateRandomNodeIdForBucket(
   return xOrNodeId(nodeId, randomDistanceForBucket);
 }
 
+function isConnectionError(e): boolean {
```
Need a comment here what this is for. What constitutes a connection error and why is it being checked here?

CMCDragonkai commented on src/nodes/NodeManager.ts:
```diff
@@ -390,13 +431,13 @@ class NodeManager {
     );
   }
 
+  // FIXME: make cancelable
```
Use `timedCancellable`. Make sure to connect it to the task deadline.

CMCDragonkai commented on src/bin/errors.ts:
```diff
@@ -49,6 +49,12 @@ class ErrorCLIPolykeyAgentProcess<T> extends ErrorCLI<T> {
   exitCode = sysexits.OSERR;
 }
 
+class ErrorCLIPolykeyAsynchronousDeadlock<T> extends ErrorCLI<T> {
```
This should not be a CLI error. This should be part of all the other errors like UnhandledPromiseRejection. We could have them in ExitHandlers.

Actually all of these can be in bin/errors.ts. But it needs to be separate from ErrorCLI, also we should change ErrorCLI to ErrorBin.

CMCDragonkai commented on src/client/service/nodesAdd.ts:
```diff
@@ -79,7 +79,6 @@ function nodesAdd({
             host,
             port,
           } as NodeAddress,
-          true,
```
What is the impact of this change?

tegefaulkes replied:
Blocking has been added back in. This change has been reverted as a result.

tegefaulkes committed:
feat: task handlers are now `timedCancellable`

tegefaulkes committed:
fix: depending on return from `updateTask` to check existence

tegefaulkes committed:
fix: replacing `private` with `protected`

tegefaulkes committed:
fix: removing `Queue.ts`

tegefaulkes committed:
fix: removing `@ready` from `TaskManager.cancelTask`

tegefaulkes committed:
docs: adding description to `isConnectionError`

tegefaulkes committed:
fix: added cancellability and blocking to `nodeManager.setNode`

tegefaulkes committed:
fix: extracted `refreshBucketsDelayJitter` into nodes utils

tegefaulkes committed:
fix: fixes to errors and adding un-recoverable error handlers

Un-recoverable errors include `ErrorBinUncaughtException`, `ErrorBinUnhandledRejection` and `ErrorBinAsynchronousDeadlock`.

#414

tegefaulkes committed:
fix: moved 'syncNodeGraph` from `NodeConnectionManager` to `NodeManager`

tegefaulkes committed:
fix: `pingNode`s inside of `garbageCollectBucket` now have timeouts separate from the overall timer

other fixes have been applied.

CMCDragonkai commented on src/nodes/NodeManager.ts:
```diff
@@ -559,183 +771,251 @@ class NodeManager {
+    block?: boolean,
+    ctx?: Partial<ContextTimed>,
+  ): PromiseCancellable<void>;
+  @ready(new nodesErrors.ErrorNodeConnectionManagerNotRunning())
```
This changes to `ErrorNodeManagerNotRunning`.

tegefaulkes replied:
This has been fixed.

CMCDragonkai commented on src/nodes/NodeConnectionManager.ts:
```diff
@@ -426,24 +419,26 @@ class NodeConnectionManager {
+   * @param ctx
    * @returns whether the target node was located in the process
    */
+  public getClosestGlobalNodes(
```
This function is not doing the operation concurrently. This never got done concurrently.

tegefaulkes replied:
Is this something we want to fix in this PR?

tegefaulkes replied:
I think I should make a new issue for this. This PR is getting too large and we should merge it sooner than later.

CMCDragonkai replied:
Depends on how easy it is to do it. And if it can use the same pattern we are using for refresh buckets, and the semaphore... etc.

tegefaulkes replied:
New issue created at #448

CMCDragonkai commented on src/nodes/NodeManager.ts:
```diff
@@ -559,183 +771,251 @@ class NodeManager {
+      } else {
+        // These are extra, so we cancel them
+        // TODO: make error
+        task.cancel(Error('TMP, cancel extra tasks'));
```
This should be it's own error class.

I would also change the async generator to speicfically deal with the first item, then all subsequent items.

tegefaulkes replied:
I would, but it's not the first task we're interested in, it's the first task that is still running, queued or scheduled. The first one could throw `ErrorTaskMissing` and that wouldn't count as a found task.

I've updated it to use a boolean flag to make it more clear. using a counter wasn't needed.

CMCDragonkai replied:
The custom error here can be a special task error that throws into the task handler for cancellation.

Can you create a error here under `src/nodes/errors.ts` for this specific purpose, then document it similar to `ErrorTaskRetry` and `ErrorTaskStop`. See `src/tasks/errors.ts`. These errors are "signals".

CMCDragonkai commented on tests/network/Proxy.test.ts:
```diff
@@ -311,16 +312,16 @@ describe(Proxy.name, () => {
       ),
     ).rejects.toThrow(networkErrors.ErrorConnectionStartTimeout);
-    timerStop(timer);
+    timer.cancel('clean up');
```
We are going to change this in a later PR, so this can be rollbacked to how it was. Integrating timers all over the codebase.

tegefaulkes replied:
I've reverted this. This change is left to the wider `timedCancellable` changes PR.

tegefaulkes committed:
fix: cleaning up ephemeral tasks when stopping `NodeManager`

tegefaulkes committed:
fix: cleaning up errors

tegefaulkes committed:
fix: small fix to `updateRefreshBucketDelay`

tegefaulkes committed:
fix: rollback of proxy changes, was out of scope for this PR

tegefaulkes committed:
fix: small fix to `garbageCollectBucket` concurrent pinging

tegefaulkes committed:
fix: updated default timeout for `NodeConnectionManager.pingNode`

tegefaulkes committed:
fix: using Symbols for cancelling tasks

tegefaulkes committed:
fix: `TaskManager` should extend the `CreateDestroyStartStop` interface

tegefaulkes committed:
fix: test wasn't overriding key-pair generation

tegefaulkes committed:
fix: `TaskManager`'s `stopProcessing` and `stopTasks` are now properly idempotent

the @ready decorator caused them to throw if ran while `taskManager` was not running. They needed to be called during incomplete startup, so I removed the decorator.

CMCDragonkai commented on src/tasks/TaskManager.ts:
```diff
@@ -245,7 +244,6 @@ class TaskManager {
    * Stop the active tasks
    * This call is idempotent
    */
-  @ready(new tasksErrors.ErrorTaskManagerNotRunning(), false, ['stopping'])
```
This is confirmed to work at any status. So no need for this.

CMCDragonkai commented on src/tasks/TaskManager.ts:
```diff
@@ -236,7 +236,6 @@ class TaskManager {
    * Stop the scheduling and queuing loop
    * This call is idempotent
    */
-  @ready(new tasksErrors.ErrorTaskManagerNotRunning(), false, ['stopping'])
```
This is confirmed to work at any status. So no need for this.

tegefaulkes committed:
tests: slightly increasing timeouts for two tests

tegefaulkes committed:
tests: general fixes for tests failing in CI

CMCDragonkai commented on src/bin/utils/ExitHandlers.ts:
```diff
@@ -83,6 +52,64 @@ class ExitHandlers {
+    // Fail fast pattern
+    process.exit();
+  };
+  protected deadlockHandler = async () => {
```
This requires a newline in between.

CMCDragonkai commented on src/bin/utils/ExitHandlers.ts:
```diff
@@ -83,6 +52,64 @@ class ExitHandlers {
+    // Fail fast pattern
+    process.exit();
+  };
+  /**
```
Newline needed.

CMCDragonkai replied:
If properties have doc block comments, separate a newline around them.

tegefaulkes committed:
syntax: formatting change for `ExitHandlers.ts`
